include/trieste/source.h

```diff
@@ -2,26 +2,38 @@
 // SPDX-License-Identifier: MIT
 #pragma once
 
+#include "intrusive_ptr.h"
+
 #include <algorithm>
 #include <cassert>
 #include <filesystem>
 #include <fstream>
 #include <iterator>
-#include <memory>
 #include <sstream>
 #include <string>
-#include <string_view>
 #include <vector>
 
 namespace trieste
 {
   class SourceDef;
   struct Location;
   class NodeDef;
-  using Source = std::shared_ptr<SourceDef>;
-  using Node = std::shared_ptr<NodeDef>;
 
-  class SourceDef
+  // Because NodeDef is an incomplete type for a while, we need to explicitly
+  // defer trying to express its refcount increment and decrement (or we get
+  // incomplete type error). The two functions that are not implemented here can
+  // be found under NodeDef's definition.
+  template<>
+  struct intrusive_refcounted_traits<NodeDef>
+  {
+    static constexpr void intrusive_inc_ref(NodeDef*);
+    static constexpr void intrusive_dec_ref(NodeDef*);
+  };
+
+  using Source = intrusive_ptr<SourceDef>;
+  using Node = intrusive_ptr<NodeDef>;
```

Comment:
A good question that I largely avoided for this PR up to this point (didn't want more compile errors because I broke something that wasn't the ptr implementation).

It can be moved one file "down" to token.h, but it stops there because there is a circular relationship between TokenDef and NodeDef.

That is probably cleaner, and it does work. I'll do that.